✨(course) add course run catalog visibility

igobranco opened the pull request:
Add catalog visibility field to course run so it's possible to hide a
course run on catalog and/or hide it completely. Closes #1502

igobranco commented:
@sampaccoud I think I need your help on this. I don't know how to fix the regression test `test_models_course_run_mark_dirty_any_field`. I'm having this error and I don't know why - `AssertionError: 0 != 1 : After refreshing from db catalog_visibility`. 

When we fix this, i think the PR would be ready to merge.

sampaccoud commented on src/richie/apps/courses/models/course.py:
```diff
@@ -660,6 +660,16 @@ class CourseRunSyncMode(models.TextChoices):
+
+    BOTH = "both", _("both - show in catalog and allow access to about page")
+    ABOUT = "about", _("about - only allow access to about page")
+    NONE = "none", _(
```
You should make the field nullable and make use of the `None` value for this.

igobranco replied:
When I change the `NONE = None` the `course_runs_sync` API enforces that the `catalog_visibility` should be `None` instead of `none`.
```
b'{"catalog_visibility":["\\"none\\" is not a valid choice."]}'
```
Then the API won't have a consistent values on the `catalog_visibility`. The supported values would be: `both`, `about` and `None`.
On Open edX Studio Advanced Settings the user inserts directly one of the values: `both`, `about` and `none`. It is not even a combo box.
So, in my opinion would be much simpler if we use the `none` string. 

Alternatively, we could use another naming, for example `hidden`.

sampaccoud replied:
I think hidden is a much better name indeed. 
none is connoted in Python and should not be used as a choice for a machine string in my opinion.

igobranco replied:
Ok. On Richie we would use `both`, `about` and `hidden` has valid values for the course run catalog visibility, including the `course_runs_sync` API.
The API clients that integrates with Open edX should dial with this difference.

sampaccoud replied:
OK we can do it in the backend I presume

igobranco replied:
Yes, I can do it on [richie_openedx_sync](https://github.com/fccn/richie-openedx-sync/blob/master/richie_openedx_sync/tasks.py#L49). If `none` then change it to `hidden`.

```python
        "catalog_visibility": 'hidden' if course.catalog_visibility == 'none' else course.catalog_visibility,
```

sampaccoud replied:
I think it is LMS specific and should be done in the backend, so probably in this method: https://github.com/openfun/richie/blob/master/src/richie/apps/courses/lms/edx.py#L77

sampaccoud replied:
Maybe don't mention the "about" page as it does not mean anything in richie?
I think we should be very explicit as it is not easy to understand:
- show on the course page and include in search results
- show on the course page and hide in search results
- hide on the course page and in search results

igobranco replied:
> I think it is LMS specific and should be done in the backend, so probably in this method: https://github.com/openfun/richie/blob/master/src/richie/apps/courses/lms/edx.py#L77

When you wrote "backend" I thought that you were talking about the LMS backend and not the richie backend.
I've changed the `clean_course_run_data` edx method implementation to also transform /adapt the data.
So make it work, I also need to change the order of evaluation on the `course_runs_sync` api. 
Previously: 
1. serializer 
2. clean data
Current: 
1. clean + transform data 
2. serializer


> Maybe don't mention the "about" page as it does not mean anything in richie? I think we should be very explicit as it is not easy to understand:
> 
> * show on the course page and include in search results
> * show on the course page and hide from search results
> * hide on the course page and from search results

I think we should show a simpler state name on the combo box. It simplifies the communication between the persons. We can change its state name or even the value of the `CourseRunCatalogVisibility` class.

Current class with reviewed messages:
```python
class CourseRunCatalogVisibility(models.TextChoices):
    """Course run catalog visibility choices."""

    BOTH = "both", _("both - show on the course page and include in search results")
    ABOUT = "about", _("about - show on the course page and hide from search results")
    HIDDEN = "hidden", _("hidden - hide on the course page and from search results")
```

Proposal class (changed from `about` to `course_page`:
```python
class CourseRunCatalogVisibility(models.TextChoices):
    """Course run catalog visibility choices."""

    BOTH = "both", _("both - show on the course page and include in search results")
    COURSE_PAGE = "course_page", _("course_page - show on the course page and hide from search results")
    HIDDEN = "hidden", _("hidden - hide on the course page and from search results")
```
@sampaccoud  What do you think about this proposal ?

sampaccoud replied:
Nice!
Could go further and change the name BOTH as well? Maybe:  "course_and_search", "course_only" and "hidden"?

igobranco replied:
> Nice! Could go further and change the name BOTH as well? Maybe: "course_and_search", "course_only" and "hidden"?

Yes, I will.

igobranco replied:
Done. 
Also adapt the edx values of the catalog visibility on the method `clean_course_run_data`.

sampaccoud commented on src/richie/apps/courses/models/course.py:
```diff
@@ -885,6 +902,16 @@ def safe_title(self):
             return None
 
+    @property
+    def is_searchable(self):
```
To be coherent with the `is_listed` field on the `Course` model, we could name this field `is_listed` as well?

igobranco replied:
In reality we don't need that method. The filter is being made directly.
https://github.com/igobranco/richie/blob/igobranco/issues/1502/course_run_catalog_visibility/src/richie/apps/search/indexers/courses.py#L580

sampaccoud replied:
ok let's remove it?

igobranco replied:
Done.

sampaccoud commented:
> @sampaccoud I think I need your help on this. I don't know how to fix the regression test `test_models_course_run_mark_dirty_any_field`. I'm having this error and I don't know why - `AssertionError: 0 != 1 : After refreshing from db catalog_visibility`.

You need to make sure that the `catalog_visibility` field changes value. I proposed a fix here: https://github.com/openfun/richie/commit/1cb0150ad2f760e82d846b12d4abcf948e767f49

igobranco commented:
> You need to make sure that the `catalog_visibility` field changes value. I proposed a fix here: [1cb0150](https://github.com/openfun/richie/commit/1cb0150ad2f760e82d846b12d4abcf948e767f49)

Thanks, it fixes the problem.

sampaccoud commented on src/richie/apps/courses/templates/courses/cms/fragment_course_runs_list.html:
```diff
@@ -4,8 +4,9 @@
     <ul class="course-detail__run-list">
         {% for run in course_runs %}
+            {% if run.is_visible_on_course_page or request.toolbar.edit_mode_active %}
             {% render_model_block run %}
```
missing indentation?

igobranco replied:
Done.

sampaccoud commented on src/richie/apps/courses/templates/courses/cms/fragment_course_run.html:
```diff
@@ -1,5 +1,6 @@
 {% load cms_tags i18n extra_tags %}{% spaceless %}
-<div class="course-detail__run-descriptions" rel="hasCourseInstance" typeof="CourseInstance">
+{% if run.is_visible_on_course_page or request.toolbar.edit_mode_active %}
+<div class="course-detail__run-descriptions course-detail__run-descriptions__{{ run.catalog_visibility|lower }}" rel="hasCourseInstance" typeof="CourseInstance">
```
missing indentation?

igobranco replied:
done

igobranco commented:
@sampaccoud  I've also included a css change to make the `hidden` runs easily distinguished on the edit mode.
My idea was to highlight only the `hidden` runs. Should we include a different style also for the `about` visibility ?

Current implementation:
![image](https://user-images.githubusercontent.com/67018/147511957-011983f1-bb36-46aa-a9c1-c2697d343d15.png)

Alternative implementation:
(only display extra information for `hidden`and `about`)

![image](https://user-images.githubusercontent.com/67018/147513588-28ee61d0-e70c-4af6-8cd2-5151b6cc0b50.png)

![image](https://user-images.githubusercontent.com/67018/147513507-33577a19-4b5d-42c1-aa35-78fbd7f3ea27.png)

sampaccoud commented:
> @sampaccoud I've also included a css change to make the `hidden` and `about` runs easily distinguished on the edit mode.

Good idea! I'm not sure about the differentiation you propose "design wise" :sweat_smile: 
I can ask the designer to propose something.

igobranco commented:
The course run api would adapt the open edx catalog visibility to the richie values, on the method `clean_course_run_data`.

Updating the previous screenshots:

### Course page (edit mode):
![image](https://user-images.githubusercontent.com/67018/148373599-5e1b5728-e95c-4007-8e38-6636bb5b1674.png)

### Course run admin screen, showing the catalog visibility combo box:
![image](https://user-images.githubusercontent.com/67018/148373660-eb56e983-aeaf-43b9-a034-294e29e569a8.png)

@sampaccoud  can you ask your designer to review the UI changes?

sampaccoud commented:
What do you think of something like this:

![sessions](https://user-images.githubusercontent.com/1427165/149318416-6bf7616f-6f00-41ca-a745-d7982c8cc5af.png)

igobranco commented:
Reviewed the styling.

Screen captures:

![image](https://user-images.githubusercontent.com/67018/149368603-061167d9-2d11-4e91-bf5f-351aa6cf50b5.png)

![image](https://user-images.githubusercontent.com/67018/149368704-603a30a1-6000-4d9b-8a0a-7351b079079f.png)

sampaccoud commented on CHANGELOG.md:
```diff
@@ -8,6 +8,12 @@ Versioning](https://semver.org/spec/v2.0.0.html).
+
+- Add catalog visibility field to course run so it's possible to hide a
+  course run on the catalog and/or hide it completely. Possible values
+  of `both`, `about` and `none`.
```
possible values should be updated

igobranco replied:
updated.

sampaccoud commented on src/richie/apps/courses/lms/edx.py:
```diff
@@ -76,11 +76,29 @@ def extract_course_number(self, data):
+                return key, change_catalog_visibility_value(value)
+            return key, value
+
+        return dict(map(adapt_fields, filter(filter_no_update_fields, data.items())))
```
missing docstrings on all added methods

igobranco replied:
They are internal methods, but ok.

sampaccoud commented on tests/apps/search/test_indexers_courses.py:
```diff
@@ -983,3 +985,103 @@ def test_indexers_courses_format_es_document_for_autocomplete(self):
+    def test_indexers_courses_get_es_documents_course_catalog_visibility_one_each(self):
+        """
+        A course with 3 runs. A run with `hidden`, another with `both` and a last one with `about`
+        on the catalog visibility should have a single run on the index.
```
you can now update the visibility values

igobranco replied:
Yes, I forgot to change the comments. I've also renamed some test method names because they included the `both` or `about` string.

sampaccoud commented on tests/apps/search/test_indexers_courses.py:
```diff
@@ -983,3 +985,103 @@ def test_indexers_courses_format_es_document_for_autocomplete(self):
+        self,
+    ):
+        """
+        A course with one run with `hidden` and another with `both` on catalog visibility should
```
ditto

igobranco replied:
done

sampaccoud commented on tests/apps/search/test_indexers_courses.py:
```diff
@@ -983,3 +985,103 @@ def test_indexers_courses_format_es_document_for_autocomplete(self):
+        self,
+    ):
+        """
+        A course run with `about` on catalog visibility should not be indexed.
```
ditto

igobranco replied:
done

sampaccoud commented on tests/apps/search/test_indexers_courses.py:
```diff
@@ -983,3 +985,103 @@ def test_indexers_courses_format_es_document_for_autocomplete(self):
+
+    def test_indexers_courses_get_es_documents_course_run_catalog_visibility_both(self):
+        """
+        A course run with `both` on catalog visibility should be indexed.
```
ditto

igobranco replied:
done

sampaccoud commented on tests/apps/courses/test_models_course_run.py:
```diff
@@ -609,15 +610,16 @@ def test_models_course_run_mark_dirty_any_field(self):
         )
+        stub = CourseRunFactory(
+            sync_mode="manual",
+            catalog_visibility=CourseRunCatalogVisibility.COURSE_ONLY,
```
why fix this field? It should not be necessary?

igobranco replied:
@sampaccoud  I don't know why, but when I remove that line, the test fails with:
```python
            self.assertEqual(
                title_obj.publisher_state,
                PUBLISHER_STATE_DIRTY,
>               msg=f"After refreshing from db {field:s}",
            )
E           AssertionError: 0 != 1 : After refreshing from db catalog_visibility
```

Can you help me with this?

sampaccoud replied:
ok I checked and sorry, it has to be modified from the default value so what you do is correct :+1:

sampaccoud commented on tests/apps/courses/test_api_course_run_sync_edx.py:
```diff
@@ -970,3 +987,129 @@ def test_api_course_run_sync_enrollment_count(self, mock_signal):
+    @override_settings(TIME_ZONE="utc")
+    def test_api_course_run_sync_catalog_visibility_none(self, mock_signal):
+        """
+        Verify that the catalog visibility is updated with `none` value
```
ditto

sampaccoud replied:
actually this is redundant with the above test?

igobranco replied:
This is already ok. I've included one test to test the adapt from open edx to Richie values. Probably I should include others cases like `both` and `about`.

igobranco replied:
I included 6 tests: The first 3 test if the catalog visibility has the Richie values, the last 3 tests the adaptation of Open edX to Richie values.
- test_api_course_run_sync_catalog_visibility_course_only
- test_api_course_run_sync_catalog_visibility_course_and_search
- test_api_course_run_sync_catalog_visibility_hidden
- test_api_course_run_sync_catalog_visibility_none
- test_api_course_run_sync_catalog_visibility_both
- test_api_course_run_sync_catalog_visibility_about

jbpenrath commented on src/richie/apps/courses/static/richie/images/catalog_visibility/hidden.svg:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<svg width="24px" height="24px" viewBox="0 0 24 24" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
+    <title>iko/prof onlysvg</title>
+    <g id="iko/prof-only" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
+        <rect id="Rectangle" x="0" y="0" width="24" height="24"></rect>
+        <path d="M1.65179764,6.64815571 C1.27396888,6.59078843 0.892686768,6.7211929 0.605822599,7.00595187 C0.318932269,7.29071084 0.154374528,7.70207552 0.154374528,8.13451147 C0.154060587,9.63235255 0.154060587,11.556694 0.154060587,11.556694 C0.154060587,15.1850516 2.86607156,18.1260938 6.21216615,18.1260938 C7.44914852,18.1260938 8.59935218,17.7239857 9.55747589,17.03413 C10.3443183,16.46796 ...
+    </g>
+</svg>
```
Here an optimized version:

```suggestion
<svg width="24" height="24" xmlns="http://www.w3.org/2000/svg"><g fill="none" fill-rule="evenodd"><path d="M0 0h24v24H0z"/><path d="M1.652 6.648a1.22 1.22 0 0 0-1.046.358 1.593 1.593 0 0 0-.452 1.129v3.422c0 3.628 2.712 6.57 6.058 6.57a5.7 5.7 0 0 0 3.345-1.093c.787-.566 1.524-.97 2.285-.97.761 0 1.498.404 2.284.97a5.701 5.701 0 0 0 3.345 1.092c3.346 0 6.059-2.94 6.059-6.57V8.135c0-.432-.165-.844-.452-1.128a1.22 1.22 0 0 0-1.046-.358c-3.375.507-6.886.675-10.19.675-3.305 0-6.815-.168-10.19-.675Zm18.73 5.602c-.327-.774-1.506-1.348-2.911-1.348-1.404 0-2.583.574-2.91 1.348.327.775 1.506 1.35 2.91 1.35 1.405 0 2.584-.575 2.91-1.35Zm-11.387-.273c-.327-.775-1.506-1.348-2.91-1.348-1.405 0-2.583.573-2.91 1.348.327.774 1.505 1.348 2.91 1.348 1.404 0 2.583-.574 2.91-1.348Z" fill="#031963"/></g></svg>
```

igobranco replied:
Ok.

And also changed the svg's color to the `indianred3` value (`#df484b`). The @sampaccoud  asked this change so we don't include a new color on the site.

jbpenrath commented on src/richie/apps/courses/static/richie/images/catalog_visibility/course_only.svg:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<svg width="24px" height="24px" viewBox="0 0 24 24" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
+    <title>iko/prof private 2svg</title>
+    <g id="iko/prof-private-2" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
+        <rect id="Rectangle" x="0" y="0" width="24" height="24"></rect>
+        <polygon id="Path" points="0 0 24 0 24 24 0 24"></polygon>
+        <path d="M19.5972081,17.4861858 L18.4850367,17.4861858 L18.0908494,17.1060766 C19.7802236,15.1351399 20.653067,12.4462192 20.1744109,9.58836114 C19.5127394,5.67464412 16.2466158,2.54930176 12.3047426,2.07064572 C6.34969832,1.33858354 1.35196616,6.35039382 2.06995022,12.3054381 C2.54860626,16.2473113 5.67394862,19.5134349 9.58766564,20.1751064 C12.4455237,20.6537625 15.1344444,19.7809191 1 ...
+    </g>
+</svg>
```
Here is an optimized version:

```suggestion
<svg width="24" height="24" xmlns="http://www.w3.org/2000/svg"><g fill="none" fill-rule="evenodd"><path d="M0 0h24v24H0z"/><path d="M0 0h24v24H0z"/><path d="M19.597 17.486h-1.112l-.394-.38a9.151 9.151 0 0 0 2.083-7.518c-.661-3.913-3.927-7.039-7.87-7.517C6.35 1.339 1.353 6.35 2.07 12.305c.479 3.942 3.604 7.208 7.518 7.87a9.151 9.151 0 0 0 7.517-2.083l.38.394v1.112l2.41 1.994a1.47 1.47 0 0 0 2.084 0l.014-.014a1.47 1.47 0 0 0 0-2.083l-2.396-2.009Zm-8.447 0a6.327 6.327 0 0 1-6.335-6.335 6.327 6.327 0 0 1 6.335-6.335 6.327 6.327 0 0 1 6.335 6.335 6.327 6.327 0 0 1-6.335 6.335Zm3.535-9.87a.99.99 0 0 0-1.414 0l-2.12 2.121-2.121-2.12a.99.99 0 0 0-1.414 0 .99.99 0 0 0 0 1.413l2.12 2.121-2.12 2.12a.99.99 0 0 0 0 1.414.99.99 0 0 0 1.414 0l2.12-2.12 2.121 2.12a.99.99 0 0 0 1.414 0 .99.99 0 0 0 0-1.413l-2.12-2.121 2.12-2.12a.99.99 0 0 0 0-1.414Z" fill="#031963" fill-rule="nonzero"/></g></svg>
```

igobranco replied:
Ditto.

jbpenrath commented on src/richie/apps/courses/templates/courses/cms/fragment_course_run.html:
```diff
@@ -1,35 +1,37 @@
-    {% endif %}
-</div>
+{% if run.is_visible_on_course_page or request.toolbar.edit_mode_active %}
+    <div class="course-detail__run-descriptions course-detail__run-descriptions__{{ run.catalog_visibility|lower }}" rel="hasCourseInstance" typeof="CourseInstance">
```
```suggestion
    <div class="course-detail__run-descriptions course-detail__run-descriptions--{{ run.catalog_visibility|lower }}" rel="hasCourseInstance" typeof="CourseInstance">
```

igobranco replied:
Changed.

jbpenrath commented on src/richie/apps/courses/templates/courses/cms/fragment_course_runs_list.html:
```diff
@@ -4,25 +4,27 @@
-            {% endrender_model_block %}
+            {% if run.is_visible_on_course_page or request.toolbar.edit_mode_active %}
+                {% render_model_block run %}
+                    <li class="{% if max_course_runs and max_course_runs < forloop.counter %}is-hidden {% endif %}course-detail__run-list__{{ run.catalog_visibility|lower }}">
```
```suggestion
                    <li class="{% if max_course_runs and max_course_runs < forloop.counter %}is-hidden {% endif %}course-detail__run-list--{{ run.catalog_visibility|lower }}">
```

igobranco replied:
Changed.

jbpenrath commented on src/frontend/scss/components/templates/courses/cms/_course_detail.scss:
```diff
@@ -131,6 +131,34 @@
     }
+
+    // Add CSS style only to edit mode
+    &__hidden {
+      h3.cms-render-model {
+        &::before {
+          background-image: r-theme-val(course-detail, run-catalog-visibility-hidden-logo);
+          content: '';
+          background-size: 1rem 1rem;
+          display: inline-block;
+          width: 1rem;
+          height: 1rem;
+          margin-left: -1rem;
+        }
+      }
+    }
+    &__course_only {
+      h3.cms-render-model {
+        &::before {
+          background-image: r-theme-val(course-detail, run-catalog-visibility-course-only-logo);
+          content: '';
+          background-size: 1rem 1rem;
+          display: inline-block;
+          width: 1rem;
+          height: 1rem;
+          margin-left: -1rem;
+        }
+      }
+    }
```
Group common styles to ease maintainability.
Furthermore to respect BEM syntax, this is modifier, so use `--` instead of `__` (e.g: ~`&__header`~ `&--header`)

```suggestion
    &--hidden,
    &--course_only {
      h3.cms-render-model::before {
        background-size: 1rem 1rem;
        content: '';
        display: inline-block;
        height: 1em;
        margin-left: -1rem;
        width: 1em;
        }
    }

    &--hidden {
      h3.cms-render-model::before {
          background-image: r-theme-val(course-detail, run-catalog-visibility-hidden-logo);
      }
    }

    &--course_only {
      h3.cms-render-model::before {
          background-image: r-theme-val(course-detail, run-catalog-visibility-course-only-logo);
      }
    }
```

igobranco replied:
Right! Changed.

jbpenrath commented on src/frontend/scss/components/templates/courses/cms/_course_detail.scss:
```diff
@@ -151,6 +179,30 @@
         margin-top: 1rem;
       }
     }
+
+    // Add CSS style only to edit mode
+    &__hidden.cms-render-model {
+      &::before {
+        background-image: r-theme-val(course-detail, run-catalog-visibility-hidden-logo);
+        content: '';
+        background-size: 1rem 1rem;
+        display: inline-block;
+        width: 1rem;
+        height: 1rem;
+        margin-left: -1.2rem;
+      }
+    }
+    &__course_only.cms-render-model {
+      &::before {
+        background-image: r-theme-val(course-detail, run-catalog-visibility-course-only-logo);
+        content: '';
+        background-size: 1rem 1rem;
+        display: inline-block;
+        width: 1rem;
+        height: 1rem;
+        margin-left: -1.2rem;
+      }
+    }
   }
```
Ditto.

Group common styles and use BEM modifier syntax `--` instead of `__`

igobranco replied:
Changed.

igobranco commented:
Screenshot: 
![image](https://user-images.githubusercontent.com/67018/150116826-0aea7693-95a0-4542-8e94-27451e2fa510.png)

igobranco commented:
@jbpenrath sorry but after I've seen what you have added on the PR #1569, I've found out that it is possible to change the svg fill color using CSS. So I've changed the new sass code and added a new theme property `course-detail` > `run-catalog-visibility-logo-color` so the color of the svg icons are inheritance from the palette.
To `mask` related properties are need to be defined neer each other. To prevent the code duplication, I've created a new sass mixin `course-detail-run-descriptions-catalog-visibility-mask`. Probably the naming is kind of strange.

With this small change, we can use our own NAU palette colors for those new svg colors more easily.

jbpenrath reviewed (approved):
LGTM ✅

igobranco commented on CHANGELOG.md:
```diff
@@ -8,6 +8,12 @@ Versioning](https://semver.org/spec/v2.0.0.html).
+
+- Add catalog visibility field to course run so it's possible to hide a
+  course run on the catalog and/or hide it completely. Possible values
+  of `course_and_search`, `course_only` and `hidden`.
```
Changed.

igobranco commented on src/richie/apps/courses/api.py:
```diff
@@ -122,16 +122,16 @@ def course_runs_sync(request, version):
 
     if serializer.is_valid() is not True:
         return Response(serializer.errors, status=400)
+    data = serializer.validated_data
```
Changed

igobranco commented on src/richie/apps/courses/factories.py:
```diff
@@ -406,6 +406,7 @@ class Meta:
     title = factory.Sequence(_("Run {:d}").format)
     resource_link = factory.Faker("uri")
     sync_mode = models.CourseRunSyncMode.SYNC_TO_PUBLIC
+    catalog_visibility = models.CourseRunCatalogVisibility.COURSE_AND_SEARCH
```
Yes, you are right. I've removed the line.
